src/platforms/hosted/windows/ftdi.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Copyright (C) 2011  Black Sphere Technologies Ltd.
+ * Copyright (C) 2018  Uwe Bonnes(bon@elektron.ikp.physik.tu-darmstadt.de)
+ * Written by Gareth McMullin <gareth@blacksphere.co.nz>
+ * Copyright (C) 2022-2023 1BitSquared <info@1bitsquared.com>
+ * Originated by Sid Price <sid@sidprice.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "general.h"
+#include "ftdi.h"
+#include "ftd2xx.h"
+#include "bmp_hosted.h"
+
+FT_HANDLE ftHandle;
+
+/*
+	The following structure is mocked by this module. FTD2XX
+	does not provide a similar structure.
+	
+	Values mocked:
+		ftdi_ctx.type	- The type of FTDI chip in the adapter
+*/
+static ftdi_context_s ftdi_ctx = {0}; // Used to fake libusb context and pass required parameters back to caller
+
+/*
+	The following array is used to map FTD2XX device type identifiers
+	to libftdi identifiers. The array is ordered by the FTD2XX
+	values, with the array entries being the libftdi values
+*/
+
+int ftdi_chip_types[] = {TYPE_AM, TYPE_BM,
+	-1, // FT_DEVICE_100AX not supported
+	-2, // Unknown type
+	TYPE_2232C, TYPE_R, TYPE_2232H, TYPE_4232H, TYPE_232H, TYPE_230X};
+
+#define number_of_ftdi_chip_types (int)(sizeof(ftdi_chip_types) / sizeof(int))
+
+#define READ_TIMEOUT  500 // Expressed in milliseconds
+#define WRITE_TIMEOUT 500
+
+struct ftdi_context *ftdi_new(void)
+{
+	return &ftdi_ctx; // JUst need to fake the structure being created
+}
+
+int ftdi_set_interface(struct ftdi_context *ftdi, enum ftdi_interface the_interface)
+{
+	(void)ftdi;
+	FT_STATUS ftStatus;
+	FT_DEVICE ftDevice;
+	DWORD device_id;
+	char serialNumber[16] = {0};
```

Comment:
ft_status
ft_device
serial_number

Please also prefer defining these where they're first used as this C89 style is bugprone (undefined behaviour use-before-init bugs can ensue)

Reply:
Confused by the statement "... prefer defining these where they are first used." I think they are defined at the begining of the function, unless you think they should be ahead of the (void)ftdi statement.

Reply:
This style of definition heralds from C89 and is considered very bugprone, while define-on-first-use eliminates the space between the declaration and the giving a value entirely as in:
```c
int ftdi_set_interface(struct ftdi_context *ftdi, enum ftdi_interface the_interface)
{
	(void)ftdi;
	//
	// FTD2XX needs a qualified serial number to open the correct device. Append
	// an interface letter to the serial number by adding the number to 'A'
	//
	char serialNumber[16] = {0};
	strcpy(serialNumber, info.serial);
	serialNumber[strlen(serialNumber)] = 'A' + (the_interface - 1);

	FT_STATUS ftStatus = FT_OpenEx(serialNumber, FT_OPEN_BY_SERIAL_NUMBER, &ftHandle);
	if (ftStatus != FT_OK)
		return 0;
	ftStatus = FT_SetTimeouts(ftHandle, READ_TIMEOUT, WRITE_TIMEOUT);
	if (ftStatus != FT_OK)
		return 0;
	FT_DEVICE ftDevice;
	DWORD device_id = 0;
	char description[64] = {'\0'};
	ftStatus = FT_GetDeviceInfo(ftHandle, &ftDevice, &device_id, serialNumber, description, NULL);
	if (ftStatus == FT_OK) {
		int device_type_index = ftDevice;
		if (device_type_index <= number_of_ftdi_chip_types) {
			ftdi_ctx.type = ftdi_chip_types[device_type_index];
		}
	}
	return 0;
}
```

This also puts the type information for the values much closer to where the value is used, but that's more a secondary benefit. C99 made this legal

Reply:
OK, my confusion, you --want-- them defined where first used

Reply:
Ah, yes! The feedback is we want to avoid C89 style and define-where-first-used, yes.